### PR TITLE
quietly release a token based metamaps api just for maps, topics, synapses, and mappings

### PR DESCRIPTION
**Pull request by Connoropolous**
notes

- [ ] need to see how the security on this actually checks out, but I think it's decent
- [ ] this is currently overriding default serializers for non-api endpoints so main site is slightly dysfunctional
- [ ] need to test 'update' and 'destroy' endpoints
- [ ] we need to actually get the permissioning on these endpoints right, which I think should join up with @devvmh pundit branch. for example, you shouldn't be able to create a mapping for a map you don't have edit rights to. you shouldn't be able to see a synapse which is connecting to a topic that is private and not yours

I did this because I want to start being able to write to metamaps from other applications. 
There's currently no UI for this, you just have to secretly get yourself a token by signing in and then using the following in the console
```
$.post('https://metamaps.cc/api/v1/tokens', {token: {
  description: 'for some application'
}})
```

then you can start making curl requests like the one I made in the postatoken.txt file

future steps might be: 
- [ ] adding scopes
- [ ] adding a UI for managing your personal access tokens, akin to githubs: https://github.com/settings/tokens

### CODE REVIEW

**Comment by Connoropolous** [comment]
this uses the same gem that loomio uses for their API to DRY up the code

:)

**Comment by Connoropolous** [comment]
@devvmh how far off is the pundit branch?

**Comment by devvmh** [comment]
@connoropolous one major next step for pundit is to clarify some permissions logic with you. Could we have a call on saturday morning (CST) to do that? 

Once I get the clarity i think it should only take 4 more hours of hacking and then testing

Clarifying may lead to some tricky decisions though, like breaking existing apis and stuff

**Comment by devvmh** [comment]
@Connoropolous looks like great stuff. pundit will definitely work nicely with this. 

I think the api examples should go into a folder.

Also, you should write tests for the api stuff! For example:

    RSpec.describe 'maps api', type: :request do
      it 'GET /maps' do
        create_list(:map, 5)
        get '/maps' # or however you get it with the api
      
        expect(response).to have_http_status(:success)
        expect(response).to match_metamaps_schema(:maps)
      end
    end

create_list is from Factory Girl, have_http_status  & get functions are in rspec 's request library. match_metamaps_schema you'd have to write yourself, but you'd specify a schema for the maps to match. That makes sure our API responses all follow consistent schemas, and then we can use those schemas in documentation. There's a 'json-schema' gem I may have already included in the codebase for defining schemas.

**Comment by Connoropolous** [comment]
Yep, I will write tests once I decided what the responses look like, I
think I'm fairly happy with them. I really like the
`active_model_serializers` gem.

On Fri, Mar 11, 2016 at 6:54 PM, Devin Howard <notifications@github.com>
wrote:

> @Connoropolous <https://github.com/Connoropolous> looks like great stuff.
> pundit will definitely work nicely with this.
>
> I think the api examples should go into a folder.
>
> Also, you should write tests for the api stuff! For example:
>
> RSpec.describe 'maps api', type: :request do
>   it 'GET /maps' do
>     create_list(:map, 5)
>     get '/maps' # or however you get it with the api
>
>     expect(response).to have_http_status(:success)
>     expect(response).to match_metamaps_schema(:maps)
>   end
> end
>
> create_list is from Factory Girl, have_http_status & get functions are in
> rspec 's request library. match_metamaps_schema you'd have to write
> yourself, but you'd specify a schema for the maps to match. That makes sure
> our API responses all follow consistent schemas, and then we can use those
> schemas in documentation. There's a 'json-schema' gem I may have already
> included in the codebase for defining schemas.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/metamaps/metamaps_gen002/pull/481#issuecomment-195240134>
> .
>



-- 
Connor Turland
~to craft a sustainable and thriving culture within and without myself~
co-creator of Metamaps.cc

**Comment by Connoropolous** [comment]
The api examples are temp code, I'll put them somewhere else ultimately, in
a much better documentation format

On Fri, Mar 11, 2016 at 6:58 PM, Connor Turland <connorturland@gmail.com>
wrote:

> Yep, I will write tests once I decided what the responses look like, I
> think I'm fairly happy with them. I really like the
> `active_model_serializers` gem.
>
> On Fri, Mar 11, 2016 at 6:54 PM, Devin Howard <notifications@github.com>
> wrote:
>
>> @Connoropolous <https://github.com/Connoropolous> looks like great
>> stuff. pundit will definitely work nicely with this.
>>
>> I think the api examples should go into a folder.
>>
>> Also, you should write tests for the api stuff! For example:
>>
>> RSpec.describe 'maps api', type: :request do
>>   it 'GET /maps' do
>>     create_list(:map, 5)
>>     get '/maps' # or however you get it with the api
>>
>>     expect(response).to have_http_status(:success)
>>     expect(response).to match_metamaps_schema(:maps)
>>   end
>> end
>>
>> create_list is from Factory Girl, have_http_status & get functions are in
>> rspec 's request library. match_metamaps_schema you'd have to write
>> yourself, but you'd specify a schema for the maps to match. That makes sure
>> our API responses all follow consistent schemas, and then we can use those
>> schemas in documentation. There's a 'json-schema' gem I may have already
>> included in the codebase for defining schemas.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/metamaps/metamaps_gen002/pull/481#issuecomment-195240134>
>> .
>>
>
>
>
> --
> Connor Turland
> ~to craft a sustainable and thriving culture within and without myself~
> co-creator of Metamaps.cc
>



-- 
Connor Turland
~to craft a sustainable and thriving culture within and without myself~
co-creator of Metamaps.cc

File: app/serializers/new_mapping_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class NewMappingSerializer < ActiveModel::Serializer
+  embed :ids, include: true
+  attributes :id,
+             :xloc,
+             :yloc,
+             :created_at,
+             :updated_at
+  has_one :user, serializer: NewUserSerializer
+  has_one :map, serializer: NewMapSerializer
+  has_one :mappable, polymorphic: true ##?
```
Comment (Connoropolous) [inline app/serializers/new_mapping_serializer.rb]: gotta figure out how to handle this polymoprhic so that it uses the right serializers in both cases...

**Comment by harlantwood** [comment]
Oh, very exciting, great work : )

**Comment by harlantwood** [comment]
@Connoropolous when I run this locally against my own server running on the latest `develop` branch:

```
$.post('/api/v1/tokens', {token: {
  description: 'myapp'
}})
```

The server 500's looking for a token policy:

```
Pundit::NotDefinedError - unable to find policy `TokenPolicy` for `Token(id: integer, token: string, description: string, user_id: integer, created_at: datetime, updated_at: datetime)`:
  pundit (1.1.0) lib/pundit/policy_finder.rb:59:in `policy!'
  pundit (1.1.0) lib/pundit.rb:112:in `policy!'
  pundit (1.1.0) lib/pundit.rb:235:in `policy'
  pundit (1.1.0) lib/pundit.rb:194:in `authorize'
  app/controllers/api/restful_controller.rb:11:in `create'
```

**Comment by harlantwood** [comment]
@Connoropolous you refer to `postatoken.txt` above, but it no longer exists.  I found it in git history tho, so for the record:

```javascript
$.post('http://localhost:3000/api/v1/tokens', {token: {
  description: 'for stuff',
  token: '1234',
  user_id: 2
}})
```

```sh
curl -X POST -d @postdata.txt http://localhost:3000/api/v1/maps --header "Authorization: Token token=fb5b3db125c94e9fb50f1e42054be856" --header "Content-Type:application/json"
```

**Comment by Connoropolous** [comment]
Aha! Thanks for those comments Harlan, we moved a little quicker than
sensible perhaps merging those things into develop, and merging various
branches and not testing all the features each time, which is of course
precisely the reason @devvmh has been writing tests and encouraging that.
We're making headway on it. Did you also find the sample json files for
posting? I have an intention to put them in some dedicated documented spot
:)

**Comment by Connoropolous** [comment]
Nvm, I see that you did find those files as well. Should prbly remove the `token` and `user_id` keys from the token object as you can't actually pass those in

**Comment by Connoropolous** [comment]
Either I'll get that token policy defined tomorrow or if you had energy for it @harlantwood that'd be nifty :) it's bed time for me, I'm in Melbourne at the moment

**Comment by Connoropolous** [comment]
Note to self we should remove 'arranged' as a validated param on the Map model

**Comment by harlantwood** [comment]
All sounds good!  I'll leave the token policy to you... When I have time though I'm very interested to try to use the new token bases API from an external application.

**Comment by Connoropolous** [comment]
Ok :) the policy thing is fixed on develop

**Comment by thoka** [comment]
Any hint, when this will go live on metamaps.cc ?

**Comment by Connoropolous** [comment]
Very soon actually! Within the next number of days most likely. Is there something that you have an interest to develop with it? or some other reason? :)

**Comment by almereyda** [comment]
@Connoropolous: He is part of our @TransforMap activities together with @AdrienLab.

@thoka You may not have been made aware of https://twitter.com/alabaeye/status/705418870027382784 which came up in discussing https://discourse.transformap.co/t/mapping-the-transformap-ecosystem/970.

**Comment by Connoropolous** [comment]
Cool @almereyda :) still curious what the specific interest is that @thoka has

**Comment by thoka** [comment]
I was looking for ways to import data from other sources and update data from other sources.

**Pull request merged**